benchmark/python/benchmark_e2e.py

```diff
@@ -0,0 +1,184 @@
+#
+# Prerequisites: 
+# 0) Install onnxruntime-genai and onnxruntime
+#
+# 1) Use builder.py to build the desired ONNX model
+#
+# 2) Run this script with the desired arguments. Run benchmark_e2e.py -h for help.
+
+import onnxruntime_genai as og
+import time
+import argparse
+
+# Use input model to generate prompt
+def generate_prompt(model, tokenizer, prompt_length) -> str:
+    prompt = "a"
+    tokens = tokenizer.encode(prompt)
+    params=og.GeneratorParams(model)
+    params.max_length = prompt_length
+    params.input_ids = tokens
+    generator=og.Generator(model, params)
+    while not generator.is_done():
+        generator.compute_logits()
+        generator.generate_next_token_top_k(50, 1.0)
+    return tokenizer.decode(generator.get_sequence(0).get_array())
+
+def save_results(results, filename):
+    import pandas as pd
+    df = pd.DataFrame(
+        results,
+        columns=[
+            "Batch Size",
+            "Prompt Length",
+            "Tokenization Throughput (tps)",
+            "Tokenization Latency (ms)",
+            "Prompt Processing Throughput (tps)",
+            "Prompt Processing Latency (ms)",
+            "Token Generation Throughput (tps)",
+            "Token Generation Latency (ms)",
+            "Sampling Throughput (tps)",
+            "Sampling Latency (ms)",
+        ],
+    )
+    df = df.transpose()  # This line swaps the rows and columns
+    df.to_csv(filename, header=False)
+    print(f"Results saved in {filename}!")
+
+def main(args):
+    # Get user arguments
+    num_repititions = args.repetitions
+    generation_length = args.generation_length
+    batch_size, prompt_length = args.batch_size, args.prompt_length
+
+    # Get tokenizer, and model
+    if args.verbose: print(f"Loading model... ")
+    model=og.Model(f'{args.input_folder}', og.DeviceType.CPU if args.device == 'cpu' else og.DeviceType.CUDA)
+    if args.verbose: print("Model loaded")
+    tokenizer = model.create_tokenizer()
+
+    # Generate prompt
+    prompt = [generate_prompt(model, tokenizer, prompt_length)] * batch_size
+    if args.verbose: print("Running warmup runs...")
+    for i in range(args.warmup):
+        if args.verbose: print(f"Running warmup repetition {i+1}...")
+        tokens = tokenizer.encode_batch(prompt)
+        params = og.GeneratorParams(model)
+        params.max_length = prompt_length + generation_length
+        params.input_ids = tokens
+        generator = og.Generator(model, params)
+        while not generator.is_done():
+            generator.compute_logits()
+            generator.generate_next_token_top_k_top_p(args.top_k, args.top_p, 1.0)
+        if args.print_model_output: print(tokenizer.decode(generator.get_sequence(0).get_array()))
+
+    tokenize_times = []
+    prompt_times = []
+    token_gen_times = []
+    sampling_times = []
+    if args.verbose: print(f"Running benchmark for batch size = {batch_size}, prompt length = {prompt_length}")
+    for i in range(num_repititions):
```

Comment:
Should we use `tqdm` so the user can see which iteration is running and how long the benchmarking will take?